In [3]:
#!pip install --upgrade transformers datasets
#!pip install --upgrade huggingface hub
#!pip install --upgrade pip
#!pip install --upgrade torch torchvision 

# Using feature embeddings to find new architectural concepts


I am going to test two models for this: DINOv2 and CLIP

#### DINOv2 Model:

In [10]:
from transformers import AutoImageProcessor, Dinov2Model
import torch, torchvision
from datasets import load_dataset

In [6]:
print(torch.__version__)
print(torchvision.__version__)

2.1.2
0.16.2


In [12]:
device = torch.device("cuda" if torch.cuda.is_available()  
                            else "mps"  if torch.backends.mps.is_available() 
                            else "cpu")
print(f"Using device: {device}")

# If you have a model, move it to the device
# model.to(device)

Using device: mps


In [13]:
#!huggingface-cli login --token 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

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /Users/ari/.cache/huggingface/token
Login successful


In [14]:
dataset = load_dataset("huggingface/cats-image")
image = dataset["test"]["image"][0]

image_processor = AutoImageProcessor.from_pretrained("facebook/dinov2-base")
model = Dinov2Model.from_pretrained("facebook/dinov2-base")

inputs = image_processor(image, return_tensors="pt")
with torch.no_grad():
    outputs = model(**inputs)

last_hidden_states = outputs.last_hidden_state
list(last_hidden_states.shape)

/Users/ari/Documents/Data_Science/3_semester/learning_from_las_vegas/LFLV/LasVegas/lib/python3.11/site-packages/datasets/load.py:1429: FutureWarning: The repository for huggingface/cats-image contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/huggingface/cats-image
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(
Generating test split: 1 examples [00:00,  2.24 examples/s]
config.json: 100%|██████████| 548/548 [00:00<00:00, 1.66MB/s]
model.safetensors: 100%|██████████| 346M/346M [00:28<00:00, 11.9MB/s] 


[1, 257, 768]

#### last hidden state shape = feature embeddings per image patch

* 1 -> input batch size
* 257 -> 256 image patches the input images are split into (one additional token for a special purpose, like classification or a start/end token). 
* 768 -> This is the dimensionality of the embeddings. Each of the 257 elements (patches or tokens) is transformed into a 768-dimensional vector. 

In [15]:
last_hidden_states

tensor([[[-2.1747, -0.4729,  1.0936,  ...,  0.2041,  1.1101,  0.1363],
         [-3.2780, -0.8269, -0.9210,  ...,  1.4415, -0.5364, -0.8757],
         [-2.9129,  1.1284, -0.7306,  ...,  0.6959, -1.8791, -2.3638],
         ...,
         [-0.5463,  1.4382, -0.2563,  ...,  0.1873, -2.9950,  0.4067],
         [-3.0848,  2.0568,  1.5137,  ...,  0.9157, -2.7059,  2.2017],
         [-0.7499,  0.0903,  1.3731,  ..., -0.2961, -2.3682, -0.1329]]])

## CLIP Model:

The Clip model outputs just one feature vector per image, because it has already aggregated the patch embeddings. Therefore this is the simpler approach. The Clip model is a little bit older than DINOv2, but also incorporates rich and complex world knowledge due to the unsupervised training approach. 

In [ ]:
%pip install git+https://github.com/openai/CLIP.git

In [ ]:
import torch
import clip
from PIL import Image

In [ ]:
# Load the model
model, preprocess = clip.load("ViT-B/32")

# Preprocess the image
image = Image.open("path_to_your_image.jpg")
image = preprocess(image).unsqueeze(0)  # Add batch dimension

# Calculate image features
with torch.no_grad():
    image_features = model.encode_image(image)

# Now you can use the image_features for your application


### openCLIP from Laion:

alternative to the older OpenAI clip model:

In [ ]:
%pip install git+https://github.com/mlfoundations/open_clip.git


In [ ]:
import torch
from PIL import Image
import open_clip

# Load the OpenCLIP model variant "H/14"
model, preprocess = open_clip.create_model_and_transforms('H/14', pretrained='laion2b_s32b_b79k')

# Preprocess the image
image_path = "path_to_your_image.jpg"
image = preprocess(Image.open(image_path)).unsqueeze(0)  # Add batch dimension

# Generate embeddings
with torch.no_grad():
    image_features = model.encode_image(image)

# image_features now contains the embeddings for your image


## Architectural epochs / styles dataset

https://www.kaggle.com/datasets/dumitrux/architectural-styles-dataset?resource=download

#### Preprocessing for DINO:
- common size with consistent aspect ratio
- normalize


In [1]:
%pip install opencv-python

  Using cached opencv_python-4.9.0.80-cp37-abi3-macosx_11_0_arm64.whl.metadata (20 kB)
Using cached opencv_python-4.9.0.80-cp37-abi3-macosx_11_0_arm64.whl (35.4 MB)
Note: you may need to restart the kernel to use updated packages.


In [ ]:
import torch
import numpy as np
from torchvision import transforms
import cv2
import os
from PIL import Image

In [ ]:
def transform_image(target_size=(224, 224)):
    """
    Create a transformation pipeline to resize and crop the image, 
    then convert it to a tensor.

    Args:
    - target_size: A tuple (height, width) for the target size.

    Returns:
    - A composed transformation function.
    """
    transform = transforms.Compose([
        transforms.Resize(target_size),  # Resize the shorter edge to target_size, maintaining aspect ratio
        transforms.CenterCrop(target_size),  # Crop the center of the image
        transforms.ToTensor(),  # Convert to tensor
    ])

    return transform

In [ ]:
# Example usage
image_folder = "path/to/your/images"
processed_images = []
transform = transform_image(target_size=(224, 224))

for image_name in os.listdir(image_folder):
    image_path = os.path.join(image_folder, image_name)
    image = Image.open(image_path).convert('RGB')  # Open image using PIL and convert to RGB
    processed_image = transform(image)  # Apply the transformation
    processed_images.append(processed_image)

# processed_images now contains tensors of your images, preprocessed and ready for the model.
